src/services/DomainModel.ts

```diff
@@ -0,0 +1,325 @@
+  */
+  unitSymbol: string; // K, M, etc.
+
+  /**
+   * The spec for the type of data produced by each reading.
+   */
+  spec: JSONValue;
+  
+
+  // todo - should we handle simple numeric values as a very common special case?
+}
+
+/**
+ * Describes a sensor reading. 
+ */
+export interface Reading {
+  /**
+   * When the reading was taken by the sensor.
+   */
+  when: DomainDate;
+
+  /**
+   * The value of the sensor reading. 
+   */
+  value: JSONValue; 
+
+}
+
+export type JSONObject = { [key in string]?: JSONValue };
+
+export type JSONValue =
+  | string
+  | number
+  | boolean
+  | JSONObject
+  | Array<JSONValue>;
+
+/**
+ * This is the model for displaying the latest values for the whole project.
+ */
+ export type ProjectReadingsSnapshot = {
+
+  /**
+   * The timestamp of the snapshot. All readings will have been taken on or before the snapshot.
+   */
+  readonly when: DomainDate;
+
+  /**
+   * The sensor hierarchy.
+   */
+  readonly project: ProjectHierarchy;
+
+  /**
+   * The readings for each sensor host. Each host is provided in the project hierarchy. 
+   * If the host does not correspond to one of the sensor hosts in the hierarchy, an exception is thrown.
+   */
+  hostReadings(sensorHost: SensorHost): SensorHostReadingsSnapshot;
+
+  /**
+   * Retrieve a reading for a sensor host, with the given name. This is intended to retrieve the known readings for a given type of sensorHost
+   * in contrast to the named cutomized readings that may be added by application developers.
+   * @param sensorHost  The sensor host the reading is present on.
+   * @param readingName The name of the reading to retrieve.
+   */
+  hostReadingByName(sensorHost: SensorHost, readingName: SensorTypeNames): Reading | undefined;
+
+}
+
+/**
+ * Historical data for gateways and nodes.
+ **/
+export type ProjectHistoricalData = { 
+  /**
+   * The period of time the historical data is taken from.
+   */ 
+  readonly period: TimePeriod;
+
+  /**
+   * The sensor hierarchy.
+   */
+  readonly project: ProjectHierarchy;
+  
+  readonly hostReadings: Map<SensorHost, SensorHostReadingsSeries>; 
+}
+
+/**
+ * Gateways and Sensors have a handful of known sensor types.
+ * These are enumerated here so that they can be queried for.
+ * @see ProjectReadingsSnapshot.hostReadingByName
+ */
+
+export const enum GatewaySensorTypeNames {
+    VOLTAGE = "gateway_voltage",
+    SIGNAL_STRENGTH = "gateway_signal_strength",
+    TEMPERATURE = "gateway_temperature",
+    LOCATION = "gateway_location",
+    LORA_SIGNAL_STRENGTH = "gateway_lora_rssi"
+}
+
+
+export const enum NodeSensorTypeNames {
+    VOLTAGE = "node_voltage",
+    LORA_SIGNAL_STRENGTH = "node_lora_rssi",
+    PIR_MOTION = "pir_motion",
+    PIR_MOTION_TOTAL = "pir_motion_total",
+    HUMIDITY = "humidity",
+    TEMPERATURE = "temperature",
+    AIR_PRESSURE = "air_pressure"
+}
+
+export const enum GatewaySensorMeasure {
+  VOLTAGE = "voltage",
+  SIGNAL_STRENGTH = "bars",
+  RSSI = "rssi",
```

Comment:
what does RSSI stand for? It's not an acronym I'm familiar with

Reply:
Received Signal Strength Indicator - it's a measure of signal strength.